File: python/blockchain.py
"""
The core blockchain.

@author GalenS <galen.scovelL@gmail.com>
"""

from time import time
from urllib.parse import urlparse

from block import Block
from node import Node
from transaction import Transaction


class Blockchain(object):
    def __init__(self):
        self.chain = []
        self.current_transactions = []
        self.nodes = []

        # Create the genesis block with no predecessors
        self.new_block(proof=100, previous_hash=1)

    def register_node(self, address):
        parsed_url = urlparse(address)
        node = Node(parsed_url.netloc)
        self.nodes.append(node)

    def new_block(self, proof, previous_hash=None):
        """
        Creates a new Block and adds it to the chain
        """
        block = Block(
            index=len(self.chain) + 1,
            timestamp=time(),
            transactions=self.current_transactions,
            proof=proof,
            previous_hash=previous_hash
        )

        # Reset current list of transactions
        self.current_transactions = []

        self.chain.append(block)

        return block

    def new_transaction(self, sender, recipient, amount):
        """
        Adds a new transaction to the list of transactions
        """
        transaction = Transaction(sender, recipient, amount)
        self.current_transactions.append(transaction)

        return self.last_block.index + 1

    @property
    def last_block(self):
        """
        Returns the last Block in the chain
        """
        return self.chain[-1]

File: scala/Blockchain.scala
package Backend

import java.util.Date

import scala.collection.mutable.ArrayBuffer


class Blockchain {
  val chain: ArrayBuffer[Block] = new ArrayBuffer[Block]()
  val currentTransactions: ArrayBuffer[Transaction] = new ArrayBuffer[Transaction]()


  def newBlock(proof: Int, previousHash: String): Block = {
    val block: Block = new Block(
      index = chain.length + 1,
      timestamp = new Date(),
      transactions = currentTransactions.toArray,
      proof = proof,
      previousHash = previousHash
    )

    currentTransactions.clear()
    chain += block

    block
  }

  def newTransaction(sender: String, recipient: String, amount: Double): Unit = {
    val transaction: Transaction = new Transaction(sender, recipient, amount)
    currentTransactions += transaction

    getLastBlock.getIndex + 1
  }

  def proofOfWork(lastProof: Int): Int = {
    var proof: Int = 1
    while (!validateProof(lastProof, proof)) {
      proof += 1
    }

    proof
  }

  def hashBlock(block: Block): String = {
    sha256Hash(block.getSortedJsonRepresentation)
  }

  private def validateProof(lastProof: Int, proof: Int): Boolean = {
    val guess = s"$lastProof$proof"
    val guessHash = sha256Hash(guess)

    guessHash.startsWith("0000")
  }

  private def sha256Hash(text: String): String = {
    String.format("%064x", new java.math.BigInteger(
        1, java.security.MessageDigest.getInstance("SHA-256").digest(text.getBytes("UTF-8"))
      )
    )
  }

  private def getLastBlock: Block = {
    chain.last
  }
}
